tests/functional/test_protocol_selection.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from botocore.model import PRIORITY_ORDERED_SUPPORTED_PROTOCOLS
+from botocore.session import Session
+from tests import ClientHTTPStubber
+from tests.functional import TEST_MODELS_DIR
+
+json_response = b"""
+{
+  "flows": [
+  ],
+  "nextToken": "eyJwYWdlIjoyfQ=="
+}
+
+"""
+
+
+def test_correct_protocol_selection():
+    # This test ensures that botocore can properly resolve a protocol and
+    # apply that to both the serializer and parser.  The logic for protocol
+    # selection is deserialized for picking the serializer and parser, so
+    # we add this test to make sure that in at least a simple circumstance
+    # where a protocol is selected that is not the same as the "protocol"
+    # trait, that we can properly create and use a serializer/parser.
+    session = Session()
+    loader = session.get_component('data_loader')
+    loader.search_paths.insert(0, TEST_MODELS_DIR)
+
+    client = session.create_client(
+        'appflow-with-more-protocols',
+        region_name='us-west-2',
+        aws_access_key_id='foo',
+        aws_secret_access_key='bar',
+    )
+
+    # This test would not be effective if the protocols were ever changed or
+    # the order of json and xml were flipped.  We need to make sure that
+    # json comes before xml and the list contains both
+    service_model_metadata = client.meta.service_model.metadata
+    assert PRIORITY_ORDERED_SUPPORTED_PROTOCOLS.index(
+        'rest-xml'
+    ) > PRIORITY_ORDERED_SUPPORTED_PROTOCOLS.index('rest-json')
```

Comment:
Is this piece relevant for the protocol selection or is it an implementation detail? If we get back `service_model_metadata['protocols'] == ['rest-json', 'rest-xml']` as True, that would imply this is also true, right?

Reply:
Well I'm not quite sure what you mean by this one.  

The purpose of this is effectively to make sure that the `protocol` field has a lower priority than the one we select.  Now that you've said that, I can change the test to check that explicitly instead of this weird coincidental check we are doing here.  But this is basically here to make sure the following can't happen:

1. We have changed the priority list so that now XML comes before JSON
2. We regress the fixes in this PR and we are using the `protocol` field for the parser again, but because XML is now higher priority, the tests don't catch it.  

I have updated this check, but if you have a better way to make sure that this test continues to be effective, I'm open to suggestions.